## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [195]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [196]:
# View the column names in the data

# There are several options that return a list of the column headers:
# df.columns
# df.columns.tolist()
# df.columns.values
# df.columns.values.tolist()
# list(df)
# [*df]

df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [197]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [198]:
# Use this space to do any additional research
# and familiarize yourself with the data.
# df.groupby('job').describe()
# df.groupby('category').describe()
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [199]:
# What three item categories had the most entries?
# df1 = df.groupby("category")
# df1.describe()
# df.info
# df["category"].value_counts().head()
df["category"].value_counts().head(3)
# df1 = df.groupby("category").size()
# print(df1.iloc[0])
# print(df1)
# df2 = df["category"].value_counts().head(1)
# print(df2)


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [200]:
# For the category with the most entries, which subcategory had the most entries?
topcategory = list(df["category"].value_counts().head(1).index)
df.loc[df["category"]==topcategory[0]]["subcategory"].value_counts().head(1)
# df.loc[df["category"]=="consumables"]["subcategory"].value_counts().head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [201]:
# Which five clients had the most entries in the data?
df["client_id"].value_counts().head(5)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [202]:
# Store the client ids of those top 5 clients in a list.
topfiveclients = list(df["client_id"].value_counts().head(5).index)
topfiveclients

[33615, 66037, 46820, 38378, 24741]

In [203]:
# How many total units (the qty column) did the client with the most entries order order?
df.loc[df["client_id"]==topfiveclients[0]]["qty"].sum()

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [204]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df["subtotal"] = df["qty"] * df["unit_price"]
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 19 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
 18  subtotal     54639 non-null  float64
dtypes: f

In [205]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df["total_weight"] = df["qty"] * df["unit_weight"]
def calc_shipping_price(x):
    if x >50:
        return x*7
    else:
        return x*10
    
df["shipping_price"] = df["total_weight"].apply(calc_shipping_price)
df[["unit_weight","qty","total_weight","shipping_price"]].head()

unit_weight  qty  total_weight  shipping_price
0         7.50  105        787.50         5512.50
1         1.49   21         31.29          312.90
2         1.68   39         65.52          458.64
3         1.23   29         35.67          356.70
4        46.43   20        928.60         6500.20

In [206]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df["tax_rate"] = 0.0925
df["total_tax"] = ((df["subtotal"] + df["shipping_price"]) * df["tax_rate"]).round(2)
df["total_price"] = df["subtotal"] + df["shipping_price"] + df["total_tax"]
df[["subtotal","shipping_price","tax_rate","total_tax","total_price"]].head()

subtotal  shipping_price  tax_rate  total_tax  total_price
0  115164.00         5512.50    0.0925   11162.58    131839.08
1     523.95          312.90    0.0925      77.41       914.26
2     527.28          458.64    0.0925      91.20      1077.12
3    1056.18          356.70    0.0925     130.69      1543.57
4    3902.00         6500.20    0.0925     962.20     11364.40

In [207]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df["total_cost"] = (df["qty"] + df["unit_cost"]) + df["shipping_price"]
df[["qty","unit_cost","shipping_price","total_cost"]].head()


qty  unit_cost  shipping_price  total_cost
0  105     762.71         5512.50     6380.21
1   21      15.09          312.90      348.99
2   39       7.86          458.64      505.50
3   29      24.85          356.70      410.55
4   20     108.17         6500.20     6628.37

In [208]:
# Create a column for the profit of each line using line cost and line price
""" Profits would also deduct the collected sales taxes.
"""

df["profit"] = df["total_price"] - (df["total_cost"] + df["total_tax"])
df["profit"].head()


0    114296.29
1       487.86
2       480.42
3      1002.33
4      3773.83
Name: profit, dtype: float64

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [209]:
# Check your work using the totals above
orders = [2742071,2173913,6128929]
for i in orders:
    total = df.loc[df["order_id"] == i]["total_price"].sum().round(2)
    print(f"Order ID {i} had a total price of ${total:,.2f}")
    # print(f"Order ID", i, "had a total price of $", df.loc[df["order_id"] == i]["total_price"].sum().round(2))




Order ID 2742071 had a total price of $152,811.89
Order ID 2173913 had a total price of $162,388.71
Order ID 6128929 had a total price of $923,441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [210]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
""" This is the top five clients by the number of times that client appears in 
the data frame, NOT the top clients by quantity.
"""

for i in topfiveclients:
    spend = df.loc[df["client_id"] == i]["total_price"].sum().round(2)
    print(f"Client ID {i} spent ${spend:,.2f}")


# topcategory = list(df["category"].value_counts().head(1).index)
# df.loc[df["category"]==topcategory[0]]["subcategory"].value_counts().head(1)


Client ID 33615 spent $8,377,308.52
Client ID 66037 spent $10,259,514.79
Client ID 46820 spent $9,743,794.36
Client ID 38378 spent $12,906,550.87
Client ID 24741 spent $82,268,892.02


In [211]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit.
""" Taxes are not considered revenues, nor are taxes included in profits.
""" 

df["revenue"] = df["total_price"] - df["total_tax"]

new_df_columns = ["qty","shipping_price","revenue","profit"]
new_client_dict = {"client_id": topfiveclients}

for e in new_df_columns:
    new_client_dict[e] = [df.loc[df["client_id"] == i][e].sum().round() for i in topfiveclients]

new_df = pd.DataFrame(new_client_dict)
new_df

client_id     qty  shipping_price     revenue      profit
0      33615   64313       1828985.0   7668017.0   5753002.0
1      66037   43018       1395152.0   9390860.0   7931181.0
2      46820   75768       1601449.0   8918805.0   7223448.0
3      38378   73667       3429455.0  11813777.0   8288952.0
4      24741  239862       5126448.0  75303334.0  69919123.0

In [212]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns.
money_columns = ["shipping_price","revenue","profit"] 


# Define a function that converts a dollar amount to millions.
def currency_format_millions(m):
    return m/1000000

# Apply the currency_format_millions function to only the money columns. 
new_df[money_columns] = new_df[money_columns].apply(currency_format_millions).round(2)
new_df

# Rename the columns to reflect the change in the money format.
new_df_clean = new_df.rename(columns={"client_id":"Client ID Number" , "qty":"Quantity Purchased" , "shipping_price":"Total Shipping Price (millions)" , "revenue":"Total Revenue (millions)" , "profit":"Total Profit (millions)"})
new_df_clean

Client ID Number  Quantity Purchased  Total Shipping Price (millions)  \
0             33615               64313                             1.83   
1             66037               43018                             1.40   
2             46820               75768                             1.60   
3             38378               73667                             3.43   
4             24741              239862                             5.13   

   Total Revenue (millions)  Total Profit (millions)  
0                      7.67                     5.75  
1                      9.39                     7.93  
2                      8.92                     7.22  
3                     11.81                     8.29  
4                     75.30                    69.92

In [213]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
new_df_sort = new_df_clean.sort_values("Total Profit (millions)",ascending=False)
new_df_sort

Client ID Number  Quantity Purchased  Total Shipping Price (millions)  \
4             24741              239862                             5.13   
3             38378               73667                             3.43   
1             66037               43018                             1.40   
2             46820               75768                             1.60   
0             33615               64313                             1.83   

   Total Revenue (millions)  Total Profit (millions)  
4                     75.30                    69.92  
3                     11.81                     8.29  
1                      9.39                     7.93  
2                      8.92                     7.22  
0                      7.67                     5.75

In [229]:
# HUZZAH!
""" The largest customer is Kendra Garrett. 
Her orders delivered $75.30 million in revenues across 239,862 items in 207 orders.
Total profit on her relationship exceeds the next four largest customer relationships, combined.
"""

print("Thank you, " + df.loc[df["client_id"] == 24741]["first"].head(1) + " " + df.loc[df["client_id"] == 24741]["last"].head(1) + "!")





416    Thank you, Kendra Garrett!
dtype: object
